### DISCUSSION TITLE
how to embed the complement server into other languages, like ts language

### DISCUSSION THREAD

**Discussion by hans000** [SPGoding/datapack-language-server#984] [General]
now i have a function like this. i hope i can use language server in other language. just like .html file can write css code.
i want to develop a vscode plugin like [styled-components](https://marketplace.visualstudio.com/items?itemName=styled-components.vscode-styled-components).
but i am not sure wheather it support

```
type Command = string

function add(value: Command) { ... }

add('say 1') // trigger completion

```

**Comment by SPGoding** [comment]
The Spyglass language server itself should work fine other than having some hard dependencies on file systems (most notably for the cache directory and the project root). The difficult part would be trying to integrate it and exchange requests with the TypeScript language service.

**Reply by MulverineX** [reply]
Prior art: https://github.com/johnsoncodehk/volar (Vue Language Server)
Another example of a benefiting project: https://github.com/TheMrZZ/sandstone (TS commands abstraction. Doesn't include types for NBT, Counts, etc due to performance concerns & TS limitations)

**Reply by MulverineX** [reply]
A built-in system for modifying other language servers, at least the TSLS, would be valuable. Spyglass Plugins would then use this system for different implementations.
